### PR TITLE
feat(ts): upstream spellsitter.nvim

### PR DESCRIPTION
**Pull request by lewis6991**
Alternative to #16650

- Unlike #16650 this runs an additional query.

### Question
* Should spell regions be defined in their own query file `spell.scm`, or should they be integrated in `highlights.scm`? This PR implements both with `spell.scm` taking higher precedence. Integrating into `highlights.scm` will allow us to run one query for both syntax highlighting and spelling.

### CODE REVIEW

**Comment by clason** [comment]
Should this be gated behind the `spell` option? I'd hate to pay a performance penalty while spelling wasn't even enabled 😬

**Comment by lewis6991** [comment]
`spell` is window-local so it has to be checked on every window update, which it does.

File: runtime/doc/treesitter.txt [hunk 1]
```diff
@@ -353,6 +353,34 @@ attribute: >
     )
 <
 
+Treesitter Spell Highlighting                            *lua-treesitter-spell*
+
+This is completely integrated with Neovim's built-in spellchecker, and
```
Comment (clason) [inline runtime/doc/treesitter.txt:358]: What is "this"? You could start with a short(!) paragraph explaining the _purpose_ of this module (and maybe how it works), then how to set spell regions, and then mention that the rest behaves like default spelling (`see |spell|`).

Comment (lewis6991) [reply runtime/doc/treesitter.txt:358]: I copied the syntax highlighting section which also begins with "this".

Comment (clason) [reply runtime/doc/treesitter.txt:358]: We can do better :) 

(vim.treesitter docs generally need some love; that's a dedicated item on the tracking issue...)

Comment (lewis6991) [reply runtime/doc/treesitter.txt]: Ok, I've had another go at re-wording.

Comment (clason) [reply runtime/doc/treesitter.txt]: Yep, that's better!

File: runtime/doc/treesitter.txt [hunk 2]
```diff
@@ -353,6 +353,39 @@ attribute: >
...
+
+The `@spell` capture is used from queries in `spell.scm` files to define spell
+regions. For example, the following query will enable spellchecking on all
+comments and string literals:
+>
+    (comment) @spell
+    (string_literal) @spell
+<
+
+Spell highlighting is integrated with Neovim's built-in spellchecker, and
```
Comment (justinmk) [inline runtime/doc/treesitter.txt:376]:
```suggestion
Spell highlighting is integrated with Nvim's built-in spellchecker, and
```

**Review by justinmk — APPROVED** [review]

**Comment by vigoux** [comment]
I just read the docs for now and I have a little question: wouldn't it make more sense to use the `(comment)` query in the last resort and us the `highlights.scm` second? IMO that would make the language support easier.

**Comment by lewis6991** [comment]
> I just read the docs for now and I have a little question: wouldn't it make more sense to use the `(comment)` query in the last resort and us the `highlights.scm` second?

If we make `highlights.scm` second, then there wouldn't be much need for a `(comment)` query.

`highlights.scm` is the last fallback because it contains many more captures and hence is a lot more inneficient than a `(comment)` query, since the spellcheck loop needs to iterate through everything.

>  IMO that would make the language support easier.

How so? Language support should be done through a `spell.scm` query file. The two fallbacks are just sane defaults for unsupported languages.

The main question, as per OP, is whether we should even have a `spell.scm`, or instead just use `highlights.scm`. If we do the latter, then we can more tightly integrate this with `highlighter.lua` and only run a single query.

**Comment by vigoux** [comment]
I think that the query resolution is a bit too complicated, that's just what I want to raise.
I did not look into the code in too much details yet too, but I think that we should really avoid relying on the `highlighter.lua`.

On the highlighter side though: mixing these two things seem really to much to me. I think that we should keep one query file per "feature" (highlights, spell, fold, etc...), and go from there. My opinion is that we should only do two mode:
1. `spell.scm` found
2. Try to guesstimate using `(comment)`.

On another note, one can use `vim.treesitter.inspect_language()` to get more information about a languge (ie the list of node types for example).

**Comment by clason** [comment]
I'm not going to comment on the technical part (although the argument "spell ~ highlight" appears sound to me and appeals to the precedent of `conceal`), but this

> It's a purely pragmatic approach to maximize UX. If we don't have the 3 points of resolution, then there will be some languages that don't get any spell-checking out of the box. This will lead to both a bad user experience, and likely additional issues on the tracker. So even though it is (slightly) more complicated, it saves us time as developers/maintainers by having it, especially for the sake of 1 line of docs + 1 line of code.

I'm not 100% happy with. I believe we need to be looking at long-term maintainability and user expectation here _first_, since it will become much harder later. The sort of issue I'm most worried about is not "why is language X not covered" but "why is language X working but not language Y, even though both don't have a `spell.scm`?"

So I would strongly suggest to either have 
1. `spell` as a predicate like `conceal` that is added to `highlights.scm` (in analogy to syntax files), or
2. a separate `spell.scm`
and _no_ fallback. (My personal preference is 2. as this is easier to grok, unless it forces significant duplication.)

The thing to remember is that we only have to be concerned about the core experience here -- which for 0.8 means Lua, Vimscript, and C at most, where it's fairly trivial to add and maintain support either way. Having a clear and unambiguous strategy will also make it _easier_ for downstream like `nvim-treesitter` to handle support requests for the up to 100+ languages they decide to maintain.

**Comment by numToStr** [comment]
One thing to note is that some language has different name for comment node. Like in Rust you have `line_comment` and `block_comment` unlike Lua/C where you only have `comment` node.

**Comment by lewis6991** [comment]
> The sort of issue I'm most worried about is not "why is language X not covered" but "why is language X working but not language Y, even though both don't have a spell.scm?"

The trade-off is that you risk a few of these sorts of issues and avoid **many** "why set spell no work?!?". Like... I added this for a reason and has massively helped in the general maintenance of https://github.com/lewis6991/spellsitter.nvim/issues, which is popular enough, so I'm not speculating here.

>  Having a clear and unambiguous strategy will also make it easier for downstream like nvim-treesitter to handle support requests for the up to 100+ languages they decide to maintain.

Downstream only has to consider the primary query method, be it `spell.scm` or `highlights.scm`.

**Comment by lewis6991** [comment]
> One thing to note is that some language has different name for comment node. Like in Rust you have `line_comment` and `block_comment` unlike Lua/C where you only have `comment` node.

This is **exactly** why we have the third resolution through `highlights.scm`.

**Comment by lewis6991** [comment]
> Also, I repeat that the situation is different here than for your plugin (and will be different afterwards), since core will not support other than the bundled languages. And if spellsitter is in core, spelling queries for other languages will be maintained by nvim-treesitter (or, rather, the individual language maintainers listed there).

Ok, as long as the additional maintenance burden on nvim-treesitter is accepted. However, I do fear this will keep users sticking to spellsitter as it will "just work" more.

**Comment by numToStr** [comment]
The doc says `@spell` capture is used from `spell.scm` but `@comment` capture is used from `highlights.scm`. This might confuse some people, and also makes me wonder what if I want to add spell check to `string` or something node. I guess something like `(string) @comment` query doesn't make sense in `highlights.scm` but can easily be justified as `(string) @spell` in `spell.scm`

So, I agree with @clason here, using only `spell.scm` with `@spell` capture.

**Comment by lewis6991** [comment]
> The doc says `@spell` capture is used from `spell.scm` but `@comment` capture is used from `highlights.scm`. This might confuse some people, and also makes me wonder what if I want to add spell check to `string` or something node. I guess something like `(string) @comment` query doesn't make sense in `highlights.scm` but can easily be justified as `(string) @spell` in `spell.scm`
> 
> So, I agree with @clason here, using only `spell.scm` with `@spell` capture.

So you simply add `(string) @spell` to `spell.scm`, we already do this for several languages. Again, the highlights.scm is just a fallback, nothing more.

**Comment by lewis6991** [comment]
Superseded by #19419

**Pull request closed without merging**